[WIN] Warning Free Code: Select

Fleshgrinder opened the pull request:
This warning was about a possible loss of data due to the downcast of `php_socket_t` to `int`. The former maps to a platform specific type, hence, it might downcast from a 64 bit integer to a 32 bit integer.

The original `nmake` warning for reference:

```
sapi\cli\php_cli_server.c(791): warning C4244: 'function': conversion from 'php_socket_t' to 'int', possible loss of data
```

I checked all callers of `sapi_cli_select`, and all of them are using `php_socket_t` properly. Hence, the change should be fine for all of internals code.

@weltling this one is for you. 🐱

weltling commented:
@Fleshgrinder nice you think about this :) That's a messy part of the Windows API on 64-bit, where SOCKET and HANDLE are 64-bit but fd is 32-bit still. For PHP, it still the target for a huge refactoring, which is not happening ATM. I at least have it on my list in the scope of async I/O probably since 2015, but timely it's still moved down the queue. 

Basically, this warning is not bad and can be ignored. Still you're right, that in the core php_select is not used with anything but sockets. The function provides the ways to work with both sockets and file descriptors, so the part for fd should be adapted. Fe see here https://github.com/php/php-src/blob/master/win32/select.c#L64 - it'll overflow now, if max_fd > INT_MAX. Also i'd still say the type in the signature should be still php_socket_t. In general this looks fine as a micro refactoring, though please check with the small corrections and possible affected places in the function itself. Fe also to think, if max_fd is passed as -1 or alike, say an invalid file descriptor. It's an exported symbol, so might be good used outside the core.

Thanks.

Fleshgrinder commented:
@weltling many thanks for your feedback.

1. Changed the signature of `php_select` from `SOCKET` to `php_socket_t` as requested.
2. Added a guard to protect against negative values by directly returning `FAILURE`.
3. I went through and checked all possibly affected variables. Now all variables within `php_select` that were `int` before are `php_socket_t` (except for the return value).
4. I fixed all places where we still have downcasts from `php_socket_t` to `int`.

However, there is one place left that I do not know how to fix:
    
```
ext\zlib\zlib_fopen_wrapper.c(141): warning C4244: 'function': conversion from 'php_socket_t' to 'int', possible loss of data
```

It is using a function/macro called `dup`, which seems to be deprecated, that comes from Windows itself and is using `int`. No clue what to do here.

I am also not sure about the possible breakages from the signature change of `php_select` for external extensions. You are more experienced with such situations than me. 😉

weltling commented:
@Fleshgrinder hmm, your latest commit doesn't look right. I was really talking about this one function php_select. SOCKET is an unsigned type, thus max_fd < 0 is always false. The looping through int might be ok, but if say -1 is passed as max_fd, it'll be casted to UINT[64]_MAX so then i++ can get > INT_MAX. Yes, a part of the solution could be to switch to looping using unsigned like you did, but then again - care about unsigned comparison and INT_MAX range and dependent codes. There are quite some subtle issues in that function, IMO it makes sense to focus on it only for now, if you're interested. 

Thanks.

weltling commented on win32/select.c:
```diff
@@ -34,18 +34,22 @@
 	struct timeval tvslice;
 	int retcode;
 
+	if (max_fd < 0) {
```
This is always false.

weltling commented on win32/select.c:
```diff
@@ -34,18 +34,22 @@
-	int handle_slot_to_fd[MAXIMUM_WAIT_OBJECTS];
-	int n_handles = 0, i;
+	php_socket_t handle_slot_to_fd[MAXIMUM_WAIT_OBJECTS];
+	php_socket_t n_handles = 0, i;
```
By having unsigned here, _get_osfhandle arg can be overflown, too. Looks like the descriptors dedicated loop might need some rework.

weltling commented on win32/select.c:
```diff
@@ -34,18 +34,22 @@
 	fd_set sock_read, sock_write, sock_except;
 	fd_set aread, awrite, aexcept;
-	int sock_max_fd = -1;
+	php_socket_t sock_max_fd = -1;
```
SOCKET is unsigned, so it'll wrap and is going to make another condition misbehave.

Fleshgrinder commented on win32/select.c:
```diff
@@ -87,7 +87,7 @@ PHPAPI int php_select(int max_fd, fd_set *rfds, fd_set *wfds, fd_set *efds, stru
 	if (n_handles == 0) {
 		/* plain sockets only - let winsock handle the whole thing */
-		return select(max_fd, rfds, wfds, efds, tv);
+		return select(NULL, rfds, wfds, efds, tv);
```
@weltling The `nfds` parameter is always ignored, according to [Microsoft’s documentation](https://msdn.microsoft.com/en-us/library/windows/desktop/ms740141(v=vs.85).aspx).

weltling replied:
Yeah, it can be ignored, doesn't look like a change required. Btw nfds is not a pointer ;)

Fleshgrinder commented:
@weltling thanks for your feedback, reverted the last commit and started anew.

We are not only facing a huge problem if we exceed `INT_MAX` in our handle loop, we are additionally constrained by the `MAXIMUM_WAIT_OBJECTS`. The loop, as it is right now, would simply access the array outside of its boundaries.

Not sure what to do in this situation. Should we ignore any further file handles? Return with a failure?

weltling commented:
@Fleshgrinder that's a good spot. Unfortunately there's no such thing similar to what `FD_SETSIZE` does. An easy fix is to indeed return FAILURE and stop processing. A proper fix would be to provide dynamic space, so then further FDs go into a new array and then do a wait loop for every array. It's not about the initial intented fix, but anyway the first variant might suffice for now given there are no uses of php_select with fd in the core, but please lets keep it outside this PR.

In general, what i was asking for - php_socket_t in the signature, that's done. And about fixing the condition `i < max_fd` as max_fd becomes ungisned and double mess on 64-bit. With both we were IMO done with your initial intention, scoped to php_select and fixed warning. Maybe i was not clear enough in the earlier posts.

To explain more - `int` is in practice sufficient to hold a `SOCKET`, even it's a mess. `INT_MAX` is huge enough in 99.999....% of cases to hold any number of mixed descriptors, created in the same process. Even if the real number of usable descriptors is likely much lower than INT_MAX, we still talk about thousands of thousands of descriptors available. The reason is simple -  the system creates descriptors incrementally and also reuses closed ones.  The number of sockets a system can create is of course another story to hold aside.  As long as the actual descriptor doesn't overflow int, there should be no issues. It is of course a certain limit, but it is a one that is almost never reached. Hence the earlier mention that the warning is not bad, but also that a much deeper refactoring is needed. For now, the remaining thing were to fix the `i < max_fd` condition, which would give the base to move forward.

Thanks.

Fleshgrinder commented:
@weltling thanks again for the feedback and explanation. I admit that I really did not get it from your first post. Did everything as requested. However, I would like to keep constant values for the first argument, to show everyone who might look at the function that these arguments are special.

weltling commented:
@Fleshgrinder oh, why closed? Otherwise - well, for 7.3 i hope to have time for some deeper I/O layer refactoring, then we'll see. I'm really inspired by what libuv does. It could require an RFC, depending on how much change comes in. I guess, some ideas to reduce efforts were a huge help.

Thanks.

Fleshgrinder commented:
@weltling oops, that was not on purpose, I deleted that branch by accident and directly restored it. I did not notice that this closed the PR.

weltling commented:
@Fleshgrinder fine then. I'm going to run some test with the patch later next week, maybe provide some follow up patch.

Thanks.

weltling commented:
Merged as 5fa1cd224b3801a99790f367feca4d98b53771b7 and some more in 7dc4d4632500b13ed4296987038967658436178b. Hopefully it can come to some serious refactoring later this year.

Thanks.
